In [1]:
import pandas as pd
import numpy as np

In [2]:
pd.options.display.float_format = '{:.2f}'.format

# Dataset loading

In [3]:
data_path = r"C:\Users\erica\Documents\Mestrado\Projetos\Advanced_DS\Ames_Housing_Project\data\raw\AmesHousing.csv"
df = pd.read_csv(data_path)
pd.set_option('display.max_columns', None)
df.head()


,Order,PID,MS SubClass,MS Zoning,Lot Frontage,Lot Area,Street,Alley,Lot Shape,Land Contour,Utilities,Lot Config,Land Slope,Neighborhood,Condition 1,Condition 2,Bldg Type,House Style,Overall Qual,Overall Cond,Year Built,Year Remod/Add,Roof Style,Roof Matl,Exterior 1st,Exterior 2nd,Mas Vnr Type,Mas Vnr Area,Exter Qual,Exter Cond,Foundation,Bsmt Qual,Bsmt Cond,Bsmt Exposure,BsmtFin Type 1,BsmtFin SF 1,BsmtFin Type 2,BsmtFin SF 2,Bsmt Unf SF,Total Bsmt SF,Heating,Heating QC,Central Air,Electrical,1st Flr SF,2nd Flr SF,Low Qual Fin SF,Gr Liv Area,Bsmt Full Bath,Bsmt Half Bath,Full Bath,Half Bath,Bedroom AbvGr,Kitchen AbvGr,Kitchen Qual,TotRms AbvGrd,Functional,Fireplaces,Fireplace Qu,Garage Type,Garage Yr Blt,Garage Finish,Garage Cars,Garage Area,Garage Qual,Garage Cond,Paved Drive,Wood Deck SF,Open Porch SF,Enclosed Porch,3Ssn Porch,Screen Porch,Pool Area,Pool QC,Fence,Misc Feature,Misc Val,Mo Sold,Yr Sold,Sale Type,Sale Condition,SalePrice
0,1,526301100,20,RL,141.00,31770,Pave,NaN,IR1,Lvl,AllPub,Corner,Gtl,NAmes,Norm,Norm,1Fam,1Story,6,5,1960,1960,Hip,CompShg,BrkFace,Plywood,Stone,112.00,TA,TA,CBlock,TA,Gd,Gd,BLQ,639.00,Unf,0.00,441.00,1080.00,GasA,Fa,Y,SBrkr,1656,0,0,1656,1.00,0.00,1,0,3,1,TA,7,Typ,2,Gd,Attchd,1960.00,Fin,2.00,528.00,TA,TA,P,210,62,0,0,0,0,NaN,NaN,NaN,0,5,2010,WD,Normal,215000
1,2,526350040,20,RH,80.00,11622,Pave,NaN,Reg,Lvl,AllPub,Inside,Gtl,NAmes,Feedr,Norm,1Fam,1Story,5,6,1961,1961,Gable,CompShg,VinylSd,VinylSd,NaN,0.00,TA,TA,CBlock,TA,TA,No,Rec,468.00,LwQ,144.00,270.00,882.00,GasA,TA,Y,SBrkr,896,0,0,896,0.00,0.00,1,0,2,1,TA,5,Typ,0,NaN,Attchd,1961.00,Unf,1.00,730.00,TA,TA,Y,140,0,0,0,120,0,NaN,MnPrv,NaN,0,6,2010,WD,Normal,105000
2,3,526351010,20,RL,81.00,14267,Pave,NaN,IR1,Lvl,AllPub,Corner,Gtl,NAmes,Norm,Norm,1Fam,1Story,6,6,1958,1958,Hip,CompShg,Wd Sdng,Wd Sdng,BrkFace,108.00,TA,TA,CBlock,TA,TA,No,ALQ,923.00,Unf,0.00,406.00,1329.00,GasA,TA,Y,SBrkr,1329,0,0,1329,0.00,0.00,1,1,3,1,Gd,6,Typ,0,NaN,Attchd,1958.00,Unf,1.00,312.00,TA,TA,Y,393,36,0,0,0,0,NaN,NaN,Gar2,12500,6,2010,WD,Normal,172000
3,4,526353030,20,RL,93.00,11160,Pave,NaN,Reg,Lvl,AllPub,Corner,Gtl,NAmes,Norm,Norm,1Fam,1Story,7,5,1968,1968,Hip,CompShg,BrkFace,BrkFace,NaN,0.00,Gd,TA,CBlock,TA,TA,No,ALQ,1065.00,Unf,0.00,1045.00,2110.00,GasA,Ex,Y,SBrkr,2110,0,0,2110,1.00,0.00,2,1,3,1,Ex,8,Typ,2,TA,Attchd,1968.00,Fin,2.00,522.00,TA,TA,Y,0,0,0,0,0,0,NaN,NaN,NaN,0,4,2010,WD,Normal,244000
4,5,527105010,60,RL,74.00,13830,Pave,NaN,IR1,Lvl,AllPub,Inside,Gtl,Gilbert,Norm,Norm,1Fam,2Story,5,5,1997,1998,Gable,CompShg,VinylSd,VinylSd,NaN,0.00,TA,TA,PConc,Gd,TA,No,GLQ,791.00,Unf,0.00,137.00,928.00,GasA,Gd,Y,SBrkr,928,701,0,1629,0.00,0.00,2,1,3,1,TA,6,Typ,1,TA,Attchd,1997.00,Fin,2.00,482.00,TA,TA,Y,212,34,0,0,0,0,NaN,MnPrv,NaN,0,3,2010,WD,Normal,189900


# Exploratory Data Analisys

## Descriptive statistics

In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 2930 entries, 0 to 2929
Data columns (total 82 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Order            2930 non-null   int64  
 1   PID              2930 non-null   int64  
 2   MS SubClass      2930 non-null   int64  
 3   MS Zoning        2930 non-null   str    
 4   Lot Frontage     2440 non-null   float64
 5   Lot Area         2930 non-null   int64  
 6   Street           2930 non-null   str    
 7   Alley            198 non-null    str    
 8   Lot Shape        2930 non-null   str    
 9   Land Contour     2930 non-null   str    
 10  Utilities        2930 non-null   str    
 11  Lot Config       2930 non-null   str    
 12  Land Slope       2930 non-null   str    
 13  Neighborhood     2930 non-null   str    
 14  Condition 1      2930 non-null   str    
 15  Condition 2      2930 non-null   str    
 16  Bldg Type        2930 non-null   str    
 17  House Style      2930 non

In [5]:
df.describe()

,Order,PID,MS SubClass,Lot Frontage,Lot Area,Overall Qual,Overall Cond,Year Built,Year Remod/Add,Mas Vnr Area,BsmtFin SF 1,BsmtFin SF 2,Bsmt Unf SF,Total Bsmt SF,1st Flr SF,2nd Flr SF,Low Qual Fin SF,Gr Liv Area,Bsmt Full Bath,Bsmt Half Bath,Full Bath,Half Bath,Bedroom AbvGr,Kitchen AbvGr,TotRms AbvGrd,Fireplaces,Garage Yr Blt,Garage Cars,Garage Area,Wood Deck SF,Open Porch SF,Enclosed Porch,3Ssn Porch,Screen Porch,Pool Area,Misc Val,Mo Sold,Yr Sold,SalePrice
count,2930.00,2930.00,2930.00,2440.00,2930.00,2930.00,2930.00,2930.00,2930.00,2907.00,2929.00,2929.00,2929.00,2929.00,2930.00,2930.00,2930.00,2930.00,2928.00,2928.00,2930.00,2930.00,2930.00,2930.00,2930.00,2930.00,2771.00,2929.00,2929.00,2930.00,2930.00,2930.00,2930.00,2930.00,2930.00,2930.00,2930.00,2930.00,2930.00
mean,1465.50,714464496.99,57.39,69.22,10147.92,6.09,5.56,1971.36,1984.27,101.90,442.63,49.72,559.26,1051.61,1159.56,335.46,4.68,1499.69,0.43,0.06,1.57,0.38,2.85,1.04,6.44,0.60,1978.13,1.77,472.82,93.75,47.53,23.01,2.59,16.00,2.24,50.64,6.22,2007.79,180796.06
std,845.96,188730844.65,42.64,23.37,7880.02,1.41,1.11,30.25,20.86,179.11,455.59,169.17,439.49,440.62,391.89,428.40,46.31,505.51,0.52,0.25,0.55,0.50,0.83,0.21,1.57,0.65,25.53,0.76,215.05,126.36,67.48,64.14,25.14,56.09,35.60,566.34,2.71,1.32,79886.69
min,1.00,526301100.00,20.00,21.00,1300.00,1.00,1.00,1872.00,1950.00,0.00,0.00,0.00,0.00,0.00,334.00,0.00,0.00,334.00,0.00,0.00,0.00,0.00,0.00,0.00,2.00,0.00,1895.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,1.00,2006.00,12789.00
25%,733.25,528477022.50,20.00,58.00,7440.25,5.00,5.00,1954.00,1965.00,0.00,0.00,0.00,219.00,793.00,876.25,0.00,0.00,1126.00,0.00,0.00,1.00,0.00,2.00,1.00,5.00,0.00,1960.00,1.00,320.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,4.00,2007.00,129500.00
50%,1465.50,535453620.00,50.00,68.00,9436.50,6.00,5.00,1973.00,1993.00,0.00,370.00,0.00,466.00,990.00,1084.00,0.00,0.00,1442.00,0.00,0.00,2.00,0.00,3.00,1.00,6.00,1.00,1979.00,2.00,480.00,0.00,27.00,0.00,0.00,0.00,0.00,0.00,6.00,2008.00,160000.00
75%,2197.75,907181097.50,70.00,80.00,11555.25,7.00,6.00,2001.00,2004.00,164.00,734.00,0.00,802.00,1302.00,1384.00,703.75,0.00,1742.75,1.00,0.00,2.00,1.00,3.00,1.00,7.00,1.00,2002.00,2.00,576.00,168.00,70.00,0.00,0.00,0.00,0.00,0.00,8.00,2009.00,213500.00
max,2930.00,1007100110.00,190.00,313.00,215245.00,10.00,9.00,2010.00,2010.00,1600.00,5644.00,1526.00,2336.00,6110.00,5095.00,2065.00,1064.00,5642.00,3.00,2.00,4.00,2.00,8.00,3.00,15.00,4.00,2207.00,5.00,1488.00,1424.00,742.00,1012.00,508.00,576.00,800.00,17000.00,12.00,2010.00,755000.00


In [6]:
print(f"Quantidade de linhas: {df.shape[0]}")
print(f"Quantidade de colunas: {df.shape[1]}")

Quantidade de linhas: 2930
Quantidade de colunas: 82


## Tabelas

In [7]:
summary_table = pd.DataFrame({
    'Num of Values': df.count(),
    '% of Missing Values': df.isnull().mean() * 100,
    'Unique Values (Categorical)': df.select_dtypes(include=['object', 'string']).nunique(),
    'Mean (Numerical)': df.select_dtypes(include='number').mean(),
    'Python Type': df.dtypes,
})

pd.set_option('display.max_rows', None)
summary_table


,Num of Values,% of Missing Values,Unique Values (Categorical),Mean (Numerical),Python Type
1st Flr SF,2930,0.00,NaN,1159.56,int64
2nd Flr SF,2930,0.00,NaN,335.46,int64
3Ssn Porch,2930,0.00,NaN,2.59,int64
Alley,198,93.24,2.00,NaN,str
Bedroom AbvGr,2930,0.00,NaN,2.85,int64
Bldg Type,2930,0.00,5.00,NaN,str
Bsmt Cond,2850,2.73,5.00,NaN,str
Bsmt Exposure,2847,2.83,4.00,NaN,str
Bsmt Full Bath,2928,0.07,NaN,0.43,float64
Bsmt Half Bath,2928,0.07,NaN,0.06,float64


### Variáveis numéricas

In [8]:
numerical_cols = df.select_dtypes(include='number').columns

In [9]:
numerical_summary = pd.DataFrame({
    'Tipo': df[numerical_cols].dtypes,
    'Média': df[numerical_cols].mean(),
    'Mediana': df[numerical_cols].median(),
    'Desvio Padrão': df[numerical_cols].std(),
    'Mínimo': df[numerical_cols].min(),
    'Máximo': df[numerical_cols].max(),
})

numerical_summary

,Tipo,Média,Mediana,Desvio Padrão,Mínimo,Máximo
Order,int64,1465.50,1465.50,845.96,1.00,2930.00
PID,int64,714464496.99,535453620.00,188730844.65,526301100.00,1007100110.00
MS SubClass,int64,57.39,50.00,42.64,20.00,190.00
Lot Frontage,float64,69.22,68.00,23.37,21.00,313.00
Lot Area,int64,10147.92,9436.50,7880.02,1300.00,215245.00
Overall Qual,int64,6.09,6.00,1.41,1.00,10.00
Overall Cond,int64,5.56,5.00,1.11,1.00,9.00
Year Built,int64,1971.36,1973.00,30.25,1872.00,2010.00
Year Remod/Add,int64,1984.27,1993.00,20.86,1950.00,2010.00
Mas Vnr Area,float64,101.90,0.00,179.11,0.00,1600.00


In [10]:
numerical_summary.sort_values(by='Desvio Padrão', ascending=False)

,Tipo,Média,Mediana,Desvio Padrão,Mínimo,Máximo
PID,int64,714464496.99,535453620.00,188730844.65,526301100.00,1007100110.00
SalePrice,int64,180796.06,160000.00,79886.69,12789.00,755000.00
Lot Area,int64,10147.92,9436.50,7880.02,1300.00,215245.00
Order,int64,1465.50,1465.50,845.96,1.00,2930.00
Misc Val,int64,50.64,0.00,566.34,0.00,17000.00
Gr Liv Area,int64,1499.69,1442.00,505.51,334.00,5642.00
BsmtFin SF 1,float64,442.63,370.00,455.59,0.00,5644.00
Total Bsmt SF,float64,1051.61,990.00,440.62,0.00,6110.00
Bsmt Unf SF,float64,559.26,466.00,439.49,0.00,2336.00
2nd Flr SF,int64,335.46,0.00,428.40,0.00,2065.00


### Variáveis Categóricas

In [11]:
catergorical_cols = df.select_dtypes(include=['object', 'string']).columns

In [12]:
categorical_summary = pd.DataFrame({
    'Valores Únicos': df[catergorical_cols].nunique(),
    'Valor Mais Frequente': df[catergorical_cols].mode().iloc[0],
    'Frequência do Valor Mais Frequente': df[catergorical_cols].apply(lambda x: x.value_counts().iloc[0]),
    'Porcentagem do Valor Mais Frequente': df[catergorical_cols].apply(lambda x: x.value_counts(normalize=True).iloc[0] * 100)
})
categorical_summary.sort_values(by='Valores Únicos', ascending=False)

,Valores Únicos,Valor Mais Frequente,Frequência do Valor Mais Frequente,Porcentagem do Valor Mais Frequente
Neighborhood,28,NAmes,443,15.12
Exterior 2nd,17,VinylSd,1015,34.64
Exterior 1st,16,VinylSd,1026,35.02
Sale Type,10,WD,2536,86.55
Condition 1,9,Norm,2522,86.08
House Style,8,1Story,1481,50.55
Functional,8,Typ,2728,93.11
Roof Matl,8,CompShg,2887,98.53
Condition 2,8,Norm,2900,98.98
MS Zoning,7,RL,2273,77.58


## Missing Values

In [13]:
varibles_with_missing_values = summary_table[summary_table['% of Missing Values'] > 0].T.columns
varibles_with_missing_values

Index(['Alley', 'Bsmt Cond', 'Bsmt Exposure', 'Bsmt Full Bath',
       'Bsmt Half Bath', 'Bsmt Qual', 'Bsmt Unf SF', 'BsmtFin SF 1',
       'BsmtFin SF 2', 'BsmtFin Type 1', 'BsmtFin Type 2', 'Electrical',
       'Fence', 'Fireplace Qu', 'Garage Area', 'Garage Cars', 'Garage Cond',
       'Garage Finish', 'Garage Qual', 'Garage Type', 'Garage Yr Blt',
       'Lot Frontage', 'Mas Vnr Area', 'Mas Vnr Type', 'Misc Feature',
       'Pool QC', 'Total Bsmt SF'],
      dtype='str')

In [14]:
for col in varibles_with_missing_values:
    print(f"{col}:\n {df[col].isnull().sum()} missing values")
    print(f"Percentage: {df[col].isnull().mean() * 100:.2f}%")
    print(f'{df[col].value_counts(dropna=False)}\n')
    print("-" * 40)

Alley:
 2732 missing values
Percentage: 93.24%
Alley
NaN     2732
Grvl     120
Pave      78
Name: count, dtype: int64

----------------------------------------
Bsmt Cond:
 80 missing values
Percentage: 2.73%
Bsmt Cond
TA     2616
Gd      122
Fa      104
NaN      80
Po        5
Ex        3
Name: count, dtype: int64

----------------------------------------
Bsmt Exposure:
 83 missing values
Percentage: 2.83%
Bsmt Exposure
No     1906
Av      418
Gd      284
Mn      239
NaN      83
Name: count, dtype: int64

----------------------------------------
Bsmt Full Bath:
 2 missing values
Percentage: 0.07%
Bsmt Full Bath
0.00    1707
1.00    1181
2.00      38
3.00       2
NaN        2
Name: count, dtype: int64

----------------------------------------
Bsmt Half Bath:
 2 missing values
Percentage: 0.07%
Bsmt Half Bath
0.00    2753
1.00     171
2.00       4
NaN        2
Name: count, dtype: int64

----------------------------------------
Bsmt Qual:
 80 missing values
Percentage: 2.73%
Bsmt Qual
TA 

In [15]:
df[varibles_with_missing_values].isnull().sum().sort_values(ascending=False)

Pool QC           2917
Misc Feature      2824
Alley             2732
Fence             2358
Mas Vnr Type      1775
Fireplace Qu      1422
Lot Frontage       490
Garage Cond        159
Garage Finish      159
Garage Qual        159
Garage Yr Blt      159
Garage Type        157
Bsmt Exposure       83
BsmtFin Type 2      81
Bsmt Cond           80
Bsmt Qual           80
BsmtFin Type 1      80
Mas Vnr Area        23
Bsmt Full Bath       2
Bsmt Half Bath       2
BsmtFin SF 1         1
Bsmt Unf SF          1
BsmtFin SF 2         1
Garage Cars          1
Garage Area          1
Electrical           1
Total Bsmt SF        1
dtype: int64

* Pool QC           2917 - Indica casas sem piscina (Structural Missing Data)
                         - Pool Area tem realação com essa variável, onde não há piscina área = 0 (imputar Nan?)

* Misc Feature      2824 - valor none proposital (Structural Missing Data)
                         - Misc Val tem relação com essa variável (imputar Nan?)

* Alley             2732 - NA significam casas sem acesso por beco (Structural Missing Data)

* Fence             2358 - casas sem cerca (Structural Missing Data)

* Mas Vnr Type      1775 - Não possuem revestimento, valor nulo proposital (Structural Missing Data)
                         - Porém, existem 7 linhas onde é nulo, mas possui área construída >0ft^2.
                         - Também existem 1745 valores que deveriam ser nulos com área = 0 - necessita imput de NaN  

* Fireplace Qu      1422 - Indica casas sem lareira (Structural Missing Data)

* Lot Frontage       490 - representa a distância linear da rua até a propriedade, provável falha no preenchimento da informação.
                         - Necessita uma estratégia de imput mais sofisticada
                         - Provável correlação com a irregularidade do terreno (?) (MAR?)

* Garage Cond        159 - Indica casas sem garagem (Structural Missing Data)
                         - Verificar justificativa da variável 'Garage Yr Blt'

* Garage Finish      159 - Indica casas sem garagem (Structural Missing Data)
                         - Verificar justificativa da variável 'Garage Yr Blt'

* Garage Qual        159 - Indica casas sem garagem (Structural Missing Data)
                         - Verificar justificativa da variável 'Garage Yr Blt'

* Garage Yr Blt      159 - Ano de contrução da garagem, como não há é nulo. (Structural Missing Data)
                         - Existem 2 linhas onde há garagem, mas não há ano de construção.
                            - uma das linhas possui área construída e vagas para carro (MCAR?)
                            - uma das linhas possui o tipo 'Detchd', mas não possui nenhuma outra informação.(MCAR?)
                            

* Garage Type        157 - Indica casas sem garagem (Structural Missing Data)
                         - Verificar justificativa da variável 'Garage Yr Blt'

* Bsmt Exposure       83 - Indica casas sem porão (Structural Missing Data)

* BsmtFin Type 2      81 - Indica casas sem porão (Structural Missing Data)

* Bsmt Cond           80 - Indica casas sem porão (Structural Missing Data)

* Bsmt Qual           80 - Indica casas sem porão (Structural Missing Data)

* BsmtFin Type 1      80 -  Indica casas sem porão (Structural Missing Data)

* Mas Vnr Area        23 - casas sem revestimento (Structural Missing Data)

* Bsmt Full Bath       2 - Essa variável indica quantidades de banheiros no porão, contudo, não existe porão nessas propriedades.
                           Isso explica o valor nulo. (Structural Missing Data)
                           - Em uma das linhas algumas características do porão está preenchida com 0, seria interessante inputar null.

* Bsmt Half Bath       2 - Mesma justificativa da variável acima 'Bsmt Full Bath'

* BsmtFin SF 1         1 - Não existe porão, justifica o valor nulo (Structural Missing Data)

* Bsmt Unf SF          1 - Mesma justificativa da variável acima 'BsmtFin SF 1'

* BsmtFin SF 2         1 - Mesma justificativa da variável acima 'BsmtFin SF 1'

* Garage Cars          1 - Corresponde a informação descrita em 'Garage Yr Blt'

* Garage Area          1 - Corresponde a informação descrita em 'Garage Yr Blt'

* Electrical           1 - Dado faltante sem explicação plausível. (MCAR?)

* Total Bsmt SF        1 - Corresponde a uma casa sem porão (Structural Missing Data)

In [16]:
# Verificando valores faltantes da coluna'pool qc' em relação a 'pool area'

df[df['Pool QC'].isnull()][['Pool Area', 'Pool QC']].value_counts(dropna=False)

Pool Area  Pool QC
0          NaN        2917
Name: count, dtype: int64

In [17]:
''' Verificando valores faltantes da coluna 'Mas Vnr Type' em relação a 'Mas Vnr Area'
 Observando que existem casos onde 'Mas Vnr Area' é maior ou igual a 0, mas 'Mas Vnr Type'
 está faltando, o que pode indicar uma inconsistência nos dados.'''

df.loc[(df['Mas Vnr Type'].isnull()) & (df['Mas Vnr Area'] >= 0), ['Mas Vnr Area', 'Mas Vnr Type']].value_counts(dropna=False)

Mas Vnr Area  Mas Vnr Type
0.00          NaN             1745
1.00          NaN                3
344.00        NaN                1
312.00        NaN                1
285.00        NaN                1
288.00        NaN                1
Name: count, dtype: int64

In [18]:
''' Verificando valores faltantes da coluna 'Misc Feature' em relação a 'Misc Val'
 Observando que existem casos onde 'Misc Val' é igual a 0, mas 'Misc Feature' está null,
 Pode indicar uma inconsistência nos dados.'''

df[df['Misc Feature'].isnull()][['Misc Feature', 'Misc Val']].value_counts(dropna=False)

Misc Feature  Misc Val
NaN           0           2824
Name: count, dtype: int64

In [ ]:
''' Verificando valores faltantes da coluna 'Fireplace Qu' em relação a 'Fireplaces'
Não há retorno de valores onde a quantidade de lareiras seja diferente de 0, o que indica que
Fireplace Qu está dentro do esperado (não existe lareira na propriedade)
'''

df[(df['Fireplace Qu'].isnull()) & (df['Fireplaces'] != 0)]

,Order,PID,MS SubClass,MS Zoning,Lot Frontage,Lot Area,Street,Alley,Lot Shape,Land Contour,Utilities,Lot Config,Land Slope,Neighborhood,Condition 1,Condition 2,Bldg Type,House Style,Overall Qual,Overall Cond,Year Built,Year Remod/Add,Roof Style,Roof Matl,Exterior 1st,Exterior 2nd,Mas Vnr Type,Mas Vnr Area,Exter Qual,Exter Cond,Foundation,Bsmt Qual,Bsmt Cond,Bsmt Exposure,BsmtFin Type 1,BsmtFin SF 1,BsmtFin Type 2,BsmtFin SF 2,Bsmt Unf SF,Total Bsmt SF,Heating,Heating QC,Central Air,Electrical,1st Flr SF,2nd Flr SF,Low Qual Fin SF,Gr Liv Area,Bsmt Full Bath,Bsmt Half Bath,Full Bath,Half Bath,Bedroom AbvGr,Kitchen AbvGr,Kitchen Qual,TotRms AbvGrd,Functional,Fireplaces,Fireplace Qu,Garage Type,Garage Yr Blt,Garage Finish,Garage Cars,Garage Area,Garage Qual,Garage Cond,Paved Drive,Wood Deck SF,Open Porch SF,Enclosed Porch,3Ssn Porch,Screen Porch,Pool Area,Pool QC,Fence,Misc Feature,Misc Val,Mo Sold,Yr Sold,Sale Type,Sale Condition,SalePrice


In [21]:
''' Verificando valores faltantes da coluna 'Lot Frontage' em relação a 'Lot Shape'
 Observando a possibilidade de que os valores faltantes ou negativos em 'Lot Frontage'
possam estar relacionados a diferentes formas de lote, o que pode indicar uma inconsistência nos dados.'''

df.loc[(df['Lot Frontage'].isnull()) | (df['Lot Frontage'] <= 0), ['Lot Frontage', 'Lot Shape']].groupby('Lot Shape').size().sort_values(ascending=False)

Lot Shape
IR1    325
Reg    132
IR2     28
IR3      5
dtype: int64

In [22]:
''' Verificando por que os valores faltantes da coluna 'Garage Type' não estão de acordo com 'Garage Yr Blt'
'''

df.loc[(df['Garage Yr Blt'].isnull()), ['Garage Type', 'Garage Yr Blt', ]].value_counts(dropna=False)

Garage Type  Garage Yr Blt
NaN          NaN              157
Detchd       NaN                2
Name: count, dtype: int64

In [23]:
''' No índice 1356, 'Garage Type' é 'Detchd', possui quantidade de carros suportados e área de garagem.
Provavelmente essa garagem existe, mas houve falha na coleta das outras informações, como o ano de construção,
acababento, qualidade e condição

Já no índice 2236, 'Garage Type' é 'Detchd', mas não possui quantidade de carros suportados e área de garagem,
o que pode indicar que a garagem não existe ou que houve erro no preenchimento da informação.'''

df.loc[(df['Garage Type'] == 'Detchd') & (df['Garage Yr Blt'].isnull())]

,Order,PID,MS SubClass,MS Zoning,Lot Frontage,Lot Area,Street,Alley,Lot Shape,Land Contour,Utilities,Lot Config,Land Slope,Neighborhood,Condition 1,Condition 2,Bldg Type,House Style,Overall Qual,Overall Cond,Year Built,Year Remod/Add,Roof Style,Roof Matl,Exterior 1st,Exterior 2nd,Mas Vnr Type,Mas Vnr Area,Exter Qual,Exter Cond,Foundation,Bsmt Qual,Bsmt Cond,Bsmt Exposure,BsmtFin Type 1,BsmtFin SF 1,BsmtFin Type 2,BsmtFin SF 2,Bsmt Unf SF,Total Bsmt SF,Heating,Heating QC,Central Air,Electrical,1st Flr SF,2nd Flr SF,Low Qual Fin SF,Gr Liv Area,Bsmt Full Bath,Bsmt Half Bath,Full Bath,Half Bath,Bedroom AbvGr,Kitchen AbvGr,Kitchen Qual,TotRms AbvGrd,Functional,Fireplaces,Fireplace Qu,Garage Type,Garage Yr Blt,Garage Finish,Garage Cars,Garage Area,Garage Qual,Garage Cond,Paved Drive,Wood Deck SF,Open Porch SF,Enclosed Porch,3Ssn Porch,Screen Porch,Pool Area,Pool QC,Fence,Misc Feature,Misc Val,Mo Sold,Yr Sold,Sale Type,Sale Condition,SalePrice
1356,1357,903426160,60,RM,57.00,8094,Pave,Grvl,Reg,Lvl,AllPub,Inside,Gtl,OldTown,Norm,Norm,2fmCon,2.5Unf,6,8,1910,1983,Gable,CompShg,Wd Sdng,Wd Sdng,NaN,0.00,TA,TA,PConc,TA,TA,Mn,Rec,196.00,Unf,0.00,1046.00,1242.00,GasA,Gd,Y,SBrkr,1242,742,0,1984,0.00,0.00,2,0,5,1,TA,8,Typ,0,NaN,Detchd,NaN,NaN,1.00,360.00,NaN,NaN,Y,64,0,180,0,0,0,NaN,MnPrv,Shed,1000,9,2008,WD,Normal,160000
2236,2237,910201180,70,RM,50.00,9060,Pave,NaN,Reg,Lvl,AllPub,Inside,Gtl,IDOTRR,Norm,Norm,1Fam,2Story,5,6,1923,1999,Gable,CompShg,Wd Sdng,Plywood,NaN,0.00,TA,TA,BrkTil,Gd,TA,No,ALQ,548.00,Unf,0.00,311.00,859.00,GasA,Ex,Y,SBrkr,942,886,0,1828,0.00,0.00,2,0,3,1,Gd,6,Typ,0,NaN,Detchd,NaN,NaN,NaN,NaN,NaN,NaN,Y,174,0,212,0,0,0,NaN,MnPrv,NaN,0,3,2007,WD,Alloca,150909


In [24]:
''' Investigando variaveis 'Bsmt' 
'''

df.loc[:, df.columns.str.startswith('Bsmt')].isnull().sum()

Bsmt Qual         80
Bsmt Cond         80
Bsmt Exposure     83
BsmtFin Type 1    80
BsmtFin SF 1       1
BsmtFin Type 2    81
BsmtFin SF 2       1
Bsmt Unf SF        1
Bsmt Full Bath     2
Bsmt Half Bath     2
dtype: int64

In [25]:


df[df['Total Bsmt SF'].isnull()]

,Order,PID,MS SubClass,MS Zoning,Lot Frontage,Lot Area,Street,Alley,Lot Shape,Land Contour,Utilities,Lot Config,Land Slope,Neighborhood,Condition 1,Condition 2,Bldg Type,House Style,Overall Qual,Overall Cond,Year Built,Year Remod/Add,Roof Style,Roof Matl,Exterior 1st,Exterior 2nd,Mas Vnr Type,Mas Vnr Area,Exter Qual,Exter Cond,Foundation,Bsmt Qual,Bsmt Cond,Bsmt Exposure,BsmtFin Type 1,BsmtFin SF 1,BsmtFin Type 2,BsmtFin SF 2,Bsmt Unf SF,Total Bsmt SF,Heating,Heating QC,Central Air,Electrical,1st Flr SF,2nd Flr SF,Low Qual Fin SF,Gr Liv Area,Bsmt Full Bath,Bsmt Half Bath,Full Bath,Half Bath,Bedroom AbvGr,Kitchen AbvGr,Kitchen Qual,TotRms AbvGrd,Functional,Fireplaces,Fireplace Qu,Garage Type,Garage Yr Blt,Garage Finish,Garage Cars,Garage Area,Garage Qual,Garage Cond,Paved Drive,Wood Deck SF,Open Porch SF,Enclosed Porch,3Ssn Porch,Screen Porch,Pool Area,Pool QC,Fence,Misc Feature,Misc Val,Mo Sold,Yr Sold,Sale Type,Sale Condition,SalePrice
1341,1342,903230120,20,RM,99.00,5940,Pave,NaN,IR1,Lvl,AllPub,FR3,Gtl,BrkSide,Feedr,Norm,1Fam,1Story,4,7,1946,1950,Gable,CompShg,MetalSd,CBlock,NaN,0.00,TA,TA,PConc,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,GasA,TA,Y,FuseA,896,0,0,896,NaN,NaN,1,0,2,1,TA,4,Typ,0,NaN,Detchd,1946.00,Unf,1.00,280.00,TA,TA,Y,0,0,0,0,0,0,NaN,MnPrv,NaN,0,4,2008,ConLD,Abnorml,79000


In [26]:
'''
Linha 444 possui 'BsmtFin SF 2' , ou seja, existe uma segunda área com acabamento do porão, mas 'BsmtFin Type 2'
está null, não há qual tipo de acabamento e está preenchido como se não houvesse porão. 
'''
df[(df['BsmtFin SF 2'] != 0) & (df['BsmtFin Type 2'].isnull())]


,Order,PID,MS SubClass,MS Zoning,Lot Frontage,Lot Area,Street,Alley,Lot Shape,Land Contour,Utilities,Lot Config,Land Slope,Neighborhood,Condition 1,Condition 2,Bldg Type,House Style,Overall Qual,Overall Cond,Year Built,Year Remod/Add,Roof Style,Roof Matl,Exterior 1st,Exterior 2nd,Mas Vnr Type,Mas Vnr Area,Exter Qual,Exter Cond,Foundation,Bsmt Qual,Bsmt Cond,Bsmt Exposure,BsmtFin Type 1,BsmtFin SF 1,BsmtFin Type 2,BsmtFin SF 2,Bsmt Unf SF,Total Bsmt SF,Heating,Heating QC,Central Air,Electrical,1st Flr SF,2nd Flr SF,Low Qual Fin SF,Gr Liv Area,Bsmt Full Bath,Bsmt Half Bath,Full Bath,Half Bath,Bedroom AbvGr,Kitchen AbvGr,Kitchen Qual,TotRms AbvGrd,Functional,Fireplaces,Fireplace Qu,Garage Type,Garage Yr Blt,Garage Finish,Garage Cars,Garage Area,Garage Qual,Garage Cond,Paved Drive,Wood Deck SF,Open Porch SF,Enclosed Porch,3Ssn Porch,Screen Porch,Pool Area,Pool QC,Fence,Misc Feature,Misc Val,Mo Sold,Yr Sold,Sale Type,Sale Condition,SalePrice
444,445,528142130,20,RL,85.00,10655,Pave,NaN,IR1,Lvl,AllPub,Inside,Gtl,NridgHt,Norm,Norm,1Fam,1Story,8,5,2003,2004,Gable,CompShg,VinylSd,VinylSd,BrkFace,296.00,Gd,TA,PConc,Gd,TA,No,GLQ,1124.00,NaN,479.00,1603.00,3206.00,GasA,Ex,Y,SBrkr,1629,0,0,1629,1.00,0.00,2,0,3,1,Gd,7,Typ,1,Gd,Attchd,2003.00,RFn,3.00,880.00,TA,TA,Y,0,0,0,0,0,0,NaN,NaN,NaN,0,10,2009,WD,Normal,284000
1341,1342,903230120,20,RM,99.00,5940,Pave,NaN,IR1,Lvl,AllPub,FR3,Gtl,BrkSide,Feedr,Norm,1Fam,1Story,4,7,1946,1950,Gable,CompShg,MetalSd,CBlock,NaN,0.00,TA,TA,PConc,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,GasA,TA,Y,FuseA,896,0,0,896,NaN,NaN,1,0,2,1,TA,4,Typ,0,NaN,Detchd,1946.00,Unf,1.00,280.00,TA,TA,Y,0,0,0,0,0,0,NaN,MnPrv,NaN,0,4,2008,ConLD,Abnorml,79000


In [27]:
'''
Não há porão nesas linhas, isso explica os dados nulos em quantidade de banheiros no porão.
'''
df[df['Bsmt Full Bath'].isnull()]

,Order,PID,MS SubClass,MS Zoning,Lot Frontage,Lot Area,Street,Alley,Lot Shape,Land Contour,Utilities,Lot Config,Land Slope,Neighborhood,Condition 1,Condition 2,Bldg Type,House Style,Overall Qual,Overall Cond,Year Built,Year Remod/Add,Roof Style,Roof Matl,Exterior 1st,Exterior 2nd,Mas Vnr Type,Mas Vnr Area,Exter Qual,Exter Cond,Foundation,Bsmt Qual,Bsmt Cond,Bsmt Exposure,BsmtFin Type 1,BsmtFin SF 1,BsmtFin Type 2,BsmtFin SF 2,Bsmt Unf SF,Total Bsmt SF,Heating,Heating QC,Central Air,Electrical,1st Flr SF,2nd Flr SF,Low Qual Fin SF,Gr Liv Area,Bsmt Full Bath,Bsmt Half Bath,Full Bath,Half Bath,Bedroom AbvGr,Kitchen AbvGr,Kitchen Qual,TotRms AbvGrd,Functional,Fireplaces,Fireplace Qu,Garage Type,Garage Yr Blt,Garage Finish,Garage Cars,Garage Area,Garage Qual,Garage Cond,Paved Drive,Wood Deck SF,Open Porch SF,Enclosed Porch,3Ssn Porch,Screen Porch,Pool Area,Pool QC,Fence,Misc Feature,Misc Val,Mo Sold,Yr Sold,Sale Type,Sale Condition,SalePrice
1341,1342,903230120,20,RM,99.00,5940,Pave,NaN,IR1,Lvl,AllPub,FR3,Gtl,BrkSide,Feedr,Norm,1Fam,1Story,4,7,1946,1950,Gable,CompShg,MetalSd,CBlock,NaN,0.00,TA,TA,PConc,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,GasA,TA,Y,FuseA,896,0,0,896,NaN,NaN,1,0,2,1,TA,4,Typ,0,NaN,Detchd,1946.00,Unf,1.00,280.00,TA,TA,Y,0,0,0,0,0,0,NaN,MnPrv,NaN,0,4,2008,ConLD,Abnorml,79000
1497,1498,908154080,20,RL,123.00,47007,Pave,NaN,IR1,Lvl,AllPub,Inside,Gtl,Edwards,Norm,Norm,1Fam,1Story,5,7,1959,1996,Gable,CompShg,Plywood,Plywood,NaN,0.00,TA,TA,Slab,NaN,NaN,NaN,NaN,0.00,NaN,0.00,0.00,0.00,GasA,TA,Y,SBrkr,3820,0,0,3820,NaN,NaN,3,1,5,1,Ex,11,Typ,2,Gd,Attchd,1959.00,Unf,2.00,624.00,TA,TA,Y,0,372,0,0,0,0,NaN,NaN,NaN,0,7,2008,WD,Normal,284700


In [28]:
df[(df['BsmtFin SF 1'].isnull()) | (df['Bsmt Unf SF'].isnull()) | (df['BsmtFin SF 2'].isnull())]

,Order,PID,MS SubClass,MS Zoning,Lot Frontage,Lot Area,Street,Alley,Lot Shape,Land Contour,Utilities,Lot Config,Land Slope,Neighborhood,Condition 1,Condition 2,Bldg Type,House Style,Overall Qual,Overall Cond,Year Built,Year Remod/Add,Roof Style,Roof Matl,Exterior 1st,Exterior 2nd,Mas Vnr Type,Mas Vnr Area,Exter Qual,Exter Cond,Foundation,Bsmt Qual,Bsmt Cond,Bsmt Exposure,BsmtFin Type 1,BsmtFin SF 1,BsmtFin Type 2,BsmtFin SF 2,Bsmt Unf SF,Total Bsmt SF,Heating,Heating QC,Central Air,Electrical,1st Flr SF,2nd Flr SF,Low Qual Fin SF,Gr Liv Area,Bsmt Full Bath,Bsmt Half Bath,Full Bath,Half Bath,Bedroom AbvGr,Kitchen AbvGr,Kitchen Qual,TotRms AbvGrd,Functional,Fireplaces,Fireplace Qu,Garage Type,Garage Yr Blt,Garage Finish,Garage Cars,Garage Area,Garage Qual,Garage Cond,Paved Drive,Wood Deck SF,Open Porch SF,Enclosed Porch,3Ssn Porch,Screen Porch,Pool Area,Pool QC,Fence,Misc Feature,Misc Val,Mo Sold,Yr Sold,Sale Type,Sale Condition,SalePrice
1341,1342,903230120,20,RM,99.00,5940,Pave,NaN,IR1,Lvl,AllPub,FR3,Gtl,BrkSide,Feedr,Norm,1Fam,1Story,4,7,1946,1950,Gable,CompShg,MetalSd,CBlock,NaN,0.00,TA,TA,PConc,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,GasA,TA,Y,FuseA,896,0,0,896,NaN,NaN,1,0,2,1,TA,4,Typ,0,NaN,Detchd,1946.00,Unf,1.00,280.00,TA,TA,Y,0,0,0,0,0,0,NaN,MnPrv,NaN,0,4,2008,ConLD,Abnorml,79000


In [29]:
df[df['Bsmt Exposure'].isnull() & (df['BsmtFin SF 1'].isnull())]

,Order,PID,MS SubClass,MS Zoning,Lot Frontage,Lot Area,Street,Alley,Lot Shape,Land Contour,Utilities,Lot Config,Land Slope,Neighborhood,Condition 1,Condition 2,Bldg Type,House Style,Overall Qual,Overall Cond,Year Built,Year Remod/Add,Roof Style,Roof Matl,Exterior 1st,Exterior 2nd,Mas Vnr Type,Mas Vnr Area,Exter Qual,Exter Cond,Foundation,Bsmt Qual,Bsmt Cond,Bsmt Exposure,BsmtFin Type 1,BsmtFin SF 1,BsmtFin Type 2,BsmtFin SF 2,Bsmt Unf SF,Total Bsmt SF,Heating,Heating QC,Central Air,Electrical,1st Flr SF,2nd Flr SF,Low Qual Fin SF,Gr Liv Area,Bsmt Full Bath,Bsmt Half Bath,Full Bath,Half Bath,Bedroom AbvGr,Kitchen AbvGr,Kitchen Qual,TotRms AbvGrd,Functional,Fireplaces,Fireplace Qu,Garage Type,Garage Yr Blt,Garage Finish,Garage Cars,Garage Area,Garage Qual,Garage Cond,Paved Drive,Wood Deck SF,Open Porch SF,Enclosed Porch,3Ssn Porch,Screen Porch,Pool Area,Pool QC,Fence,Misc Feature,Misc Val,Mo Sold,Yr Sold,Sale Type,Sale Condition,SalePrice
1341,1342,903230120,20,RM,99.00,5940,Pave,NaN,IR1,Lvl,AllPub,FR3,Gtl,BrkSide,Feedr,Norm,1Fam,1Story,4,7,1946,1950,Gable,CompShg,MetalSd,CBlock,NaN,0.00,TA,TA,PConc,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,GasA,TA,Y,FuseA,896,0,0,896,NaN,NaN,1,0,2,1,TA,4,Typ,0,NaN,Detchd,1946.00,Unf,1.00,280.00,TA,TA,Y,0,0,0,0,0,0,NaN,MnPrv,NaN,0,4,2008,ConLD,Abnorml,79000


In [31]:
df[df['Electrical'].isnull()]

,Order,PID,MS SubClass,MS Zoning,Lot Frontage,Lot Area,Street,Alley,Lot Shape,Land Contour,Utilities,Lot Config,Land Slope,Neighborhood,Condition 1,Condition 2,Bldg Type,House Style,Overall Qual,Overall Cond,Year Built,Year Remod/Add,Roof Style,Roof Matl,Exterior 1st,Exterior 2nd,Mas Vnr Type,Mas Vnr Area,Exter Qual,Exter Cond,Foundation,Bsmt Qual,Bsmt Cond,Bsmt Exposure,BsmtFin Type 1,BsmtFin SF 1,BsmtFin Type 2,BsmtFin SF 2,Bsmt Unf SF,Total Bsmt SF,Heating,Heating QC,Central Air,Electrical,1st Flr SF,2nd Flr SF,Low Qual Fin SF,Gr Liv Area,Bsmt Full Bath,Bsmt Half Bath,Full Bath,Half Bath,Bedroom AbvGr,Kitchen AbvGr,Kitchen Qual,TotRms AbvGrd,Functional,Fireplaces,Fireplace Qu,Garage Type,Garage Yr Blt,Garage Finish,Garage Cars,Garage Area,Garage Qual,Garage Cond,Paved Drive,Wood Deck SF,Open Porch SF,Enclosed Porch,3Ssn Porch,Screen Porch,Pool Area,Pool QC,Fence,Misc Feature,Misc Val,Mo Sold,Yr Sold,Sale Type,Sale Condition,SalePrice
1577,1578,916386080,80,RL,73.00,9735,Pave,NaN,Reg,Lvl,AllPub,Inside,Gtl,Timber,Norm,Norm,1Fam,SLvl,5,5,2006,2007,Gable,CompShg,VinylSd,VinylSd,NaN,0.00,TA,TA,PConc,Gd,TA,No,Unf,0.00,Unf,0.00,384.00,384.00,GasA,Gd,Y,NaN,754,640,0,1394,0.00,0.00,2,1,3,1,Gd,7,Typ,0,NaN,BuiltIn,2007.00,Fin,2.00,400.00,TA,TA,Y,100,0,0,0,0,0,NaN,NaN,NaN,0,5,2008,WD,Normal,167500
